rustual-boy-middleware/src/lib.rs

```diff
@@ -0,0 +1,82 @@
+extern crate rustual_boy_core;
+
+mod color;
+
+pub use color::Color;
+
+use rustual_boy_core::video_frame_sink::VideoFrameSink;
+use rustual_boy_core::vip::{DISPLAY_RESOLUTION_X, DISPLAY_RESOLUTION_Y};
+
+const DISPLAY_PIXELS: usize = DISPLAY_RESOLUTION_X * DISPLAY_RESOLUTION_Y;
+
+/// A VideoFrameSink for the Rustual Boy core that collapses the left/right
+/// anaglyph channels in to a single buffer.
+pub struct AnaglyphFrameSink {
+    /// Color of the left channel
+    left_color: Color,
+    /// Color of the right channel
+    right_color: Color,
+
+    inner: Option<(Box<[u8]>, Box<[u8]>)>,
+}
+
+impl AnaglyphFrameSink {
+    /// Create a new AnaglyphFrameSink which will use the provided colors for
+    /// the left and right channels
+    pub fn new(left_color: Color, right_color: Color) -> AnaglyphFrameSink {
+        AnaglyphFrameSink {
+            left_color: left_color,
+            right_color: right_color,
+            inner: None
+        }
+    }
+
+    pub fn update_availible(&self) -> bool {
```

Comment:
I think rather than having this struct contain a single frame, it should maybe rather take in another `VideoFrameSink` (or some other sink that just has one output buffer in this case), and whenever it receives a frame, it will populate the underlying sink. A sink that just either does or doesn't contain a buffer isn't a great design for a sink (unless you _know_ you only want the latest value going into the sink); the reason I was doing that in the top-level code was because we were always just going to update the window's contents with the latest buffer. However, if we had a `VideoFrameSink` that were producing an image stream for a video encoder, for example, we'd want to know about _all_ frames being produced by the emulator; not just the latest.

This design would move the `Option`-related stuff back into another inner sink in the top-level modules (where it belongs I think), and would be used something like this:

```
// setup
let frame_sink = SomeTopLevelSinkWithOptionFrame::new();
let mut anaglyph_frame_sink = AnaglyphFrameSink::new(frame_sink, left_color, right_color); // Take ownership so this sink can mutate the inner one properly

...

// After a frame in rustual-boy-cli, for example:
let frame_sink = anaglyph_frame_sink.to_inner(); // Pull out inner sink so we can see if there's a frame and take it out
```

Does this make sense? The goal here is to make the structs in this crate as general-purpose as possible with regards to data routing and let the emulators themselves decide what to to with them. It's also quite possible that `VideoFrameSink` might actually be a more general `Sink` with a left/right buffer, which could make it even easier to make this more general.